src/components/MoneyInput/index.tsx

```diff
@@ -120,6 +128,29 @@ export const MoneyInput: React.FC<MoneyInputProps> = ({
   const { isMobile } = useDeviceInfo()
   const inputId = `money-input-${useId()}`
 
+  /*
+    This will make sure that when you define a maximum
+    amount, we're reseting the amount to the maxiumum
+    amount if the user is overflowing this. We're
+    only doing this for non-controlled inputs, so
+    if you're using a controlled input you should
+    handle this yourself.
+  */
+  useEffect(() => {
+    if (
+      !isAmountControlled &&
+      hasMaximumAmount &&
+      parseAmount(amount) > maximumAmount
+    ) {
+      setAmount(maximumAmount / 100)
+    } else {
+      onChange({
+        currency,
+        amount: isAmountControlled ? amount : parseAmount(amount),
+      })
+    }
+  }, [amount])
```

Comment:
There are a lot of dependencies missing on the dependency array. Also, the `onChange` being called a lot in the `useEffect` doesn't seem right. Might be nice to use a `useCallback` here like this:

```
  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const amount = event.target.value
      if (
        !isAmountControlled &&
        hasMaximumAmount &&
        parseAmount(amount) > maximumAmount
      ) {
        setAmount(maximumAmount / 100)
      } else {
        setAmount(amount)
      }
    },
    [hasMaximumAmount, isAmountControlled, maximumAmount]
  )
  ```
  
  Then you add this `handleChange` to the onChange prop and it will already generate one less render.

Reply:
I'll check this out :)